actionpack/lib/action_controller/metal/rate_limiting.rb

```diff
@@ -44,19 +49,73 @@ module ClassMethods
       #       RATE_LIMIT_STORE = ActiveSupport::Cache::RedisCacheStore.new(url: ENV["REDIS_URL"])
       #       rate_limit to: 10, within: 3.minutes, store: RATE_LIMIT_STORE
       #     end
-      def rate_limit(to:, within:, by: -> { request.remote_ip }, with: -> { head :too_many_requests }, store: cache_store, **options)
-        before_action -> { rate_limiting(to: to, within: within, by: by, with: with, store: store) }, **options
+      #
+      #     class SessionsController < ApplicationController
+      #       rate_limit to: 10, within: 3.minutes, only: :create, strategy :sliding_window
+      #     end
+      def rate_limit(to:, within:, by: -> { request.remote_ip }, with: -> { head :too_many_requests }, store: cache_store, strategy: :fixed_window, **options)
+        before_action -> { rate_limiting(to: to, within: within, by: by, with: with, store: store, strategy: strategy) }, **options
+      end
+    end
+
+    class BaseStrategy
+      attr_reader :to, :within, :store
+
+      def initialize(to:, within:, store:)
+        @to = to
+        @within = within
+        @store = store
+      end
+
+      def increment_and_check(key:)
+        raise NotImplementedError, "Subclasses must implement `increment_and_check` method"
       end
     end
 
     private
-      def rate_limiting(to:, within:, by:, with:, store:)
-        count = store.increment("rate-limit:#{controller_path}:#{instance_exec(&by)}", 1, expires_in: within)
-        if count && count > to
+      def rate_limiting(to:, within:, by:, with:, store:, strategy:)
+        key = "rate-limit:#{controller_path}:#{instance_exec(&by)}"
+        strategy_class = if strategy.is_a?(Class)
+          strategy
+        elsif STRATEGY_CLASSES.key?(strategy)
+          STRATEGY_CLASSES[strategy]
+        else
+          raise ArgumentError, "Unknown rate limiting strategy: #{strategy}"
+        end
+
+        limiter = strategy_class.new(to: to, within: within, store: store)
+
+        if limiter.increment_and_check(key: key)
           ActiveSupport::Notifications.instrument("rate_limit.action_controller", request: request) do
             instance_exec(&with)
           end
         end
       end
+
+      class FixedWindowStrategy < BaseStrategy
+        def increment_and_check(key:)
+          count = store.increment(key, 1, expires_in: within)
+          count && count > to
+        end
+      end
+
+      class SlidingWindowStrategy < BaseStrategy
+        def increment_and_check(key:)
+          current_time = Time.now.to_f
+          bucket = (store.read(key) || []).select { |timestamp| timestamp > current_time - within }
+          bucket.push(current_time)
+          store.write(key, bucket, expires_in: within)
```

Comment:
This is extremely racy. I don't think you can realistically implement that in a backend independent way, hence why it's likely better to leave it to third party gems.